nifi-nar-bundles/nifi-graph-bundle/nifi-other-graph-services/src/main/java/org/apache/nifi/graph/TinkerpopClientService.java

```diff
@@ -0,0 +1,538 @@
+            EXTRA_RESOURCE,
+            EXTENSION_CLASSES,
+            CONTACT_POINTS,
+            PORT,
+            PATH,
+            TRAVERSAL_SOURCE_NAME,
+            USER_NAME,
+            PASSWORD,
+            SSL_CONTEXT_SERVICE
+    ));
+
+    private ScriptEngineManager MANAGER = new ScriptEngineManager();
+    private ScriptEngine engine;
+    private Map<String, CompiledScript> compiledCode;
+    protected Cluster cluster;
+    private String traversalSourceName;
+    private GraphTraversalSource traversalSource;
+    private boolean scriptSubmission = true;
+    boolean usesSSL;
+    protected String transitUrl;
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) throws InitializationException {
+        loadClasses(context);
+        compiledCode = new ConcurrentHashMap<>();
+        engine = MANAGER.getEngineByName("groovy");
+
+        if (context.getProperty(TRAVERSAL_SOURCE_NAME).isSet()) {
+            traversalSourceName = context.getProperty(TRAVERSAL_SOURCE_NAME).evaluateAttributeExpressions()
+                    .getValue();
+        }
+
+        scriptSubmission = context.getProperty(SUBMISSION_TYPE).getValue().equals(SCRIPT_SUBMISSION.getValue());
+
+        cluster = buildCluster(context);
+    }
+
+    @OnDisabled
+    public void shutdown() {
+        try {
+            compiledCode = null;
+            engine = null;
+            if (traversalSource != null) {
+                traversalSource.close();
+            }
+        } catch (Exception e) {
+            throw new ProcessException(e);
+        } finally {
+            if (cluster != null) {
+                cluster.close();
+            }
+            cluster = null;
+            traversalSource = null;
+        }
+    }
+
+    @Override
+    public Map<String, String> executeQuery(String s, Map<String, Object> map, GraphQueryResultCallback graphQueryResultCallback) {
+        try {
+            if (scriptSubmission) {
+                return scriptSubmission(s, map, graphQueryResultCallback);
+            } else {
+                return bytecodeSubmission(s, map, graphQueryResultCallback);
+            }
+        } catch (Exception ex) {
+            throw new ProcessException(ex);
+        }
+    }
+
+    @Override
+    public String getTransitUrl() {
+        return this.transitUrl;
+    }
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return DESCRIPTORS;
+    }
+
+    @Override
+    public Collection<ValidationResult> customValidate(ValidationContext context) {
+        Collection<ValidationResult> results = new ArrayList<>();
+        boolean jarsIsSet = !StringUtils.isEmpty(context.getProperty(EXTRA_RESOURCE).getValue());
+        boolean clzIsSet = !StringUtils.isEmpty(context.getProperty(EXTENSION_CLASSES).getValue());
+
+        if (jarsIsSet && clzIsSet) {
+            try {
+                final ClassLoader loader = this.getClass().getClassLoader();
```

Comment:
I don't think this works on 2.x and Java 17. These are the changes I had to make to another project.

<img width="827" alt="image" src="https://github.com/apache/nifi/assets/108184/948aa7a9-a0c6-4fcb-9017-158cba19f35b">

@exceptionfactory thoughts?

Reply:
I had tested this code using JDK17, and it loaded the custom classes just fine, however if this is the more acceptable use case, I can switch it over and test.

Reply:
Tested and this works as expected.

Reply:
So one odd thing, in all my testing, it doesnt matter if I set EXTENSION_CLASSES to be any specific classes I need, they seem to be available as long as I have the classes available inside of the Jar(s) present in EXTENSION_CLASSES. I have included unit test that proves this. Any thoughts on just removing the EXTENSION_CLASSES? I could imaging keeping it as we would need functionality such as this if we are to move away from Groovy entirely within NiFi.